### ISSUE TITLE
Should goSam have more configurable behavior?

### ISSUE THREAD

**Issue by eyedeekay** [cryptix/goSam#5]
I've really been loving writing my applications with goSam, thank's very much for writing the library. As I use it, though, I sometimes find myself wishing for more flexibility(Like with the default host:port pull request I just sent in). I think I could add in more configurable behavior for applications to leverage in much the same way, to enable things like variable hop/tunnel counts and other features of i2p. I'd like to know if this is something you'd like added to goSam(I could do it)? Also if you are interested, I'd like to briefly discuss defaults for any new settings.

**Comment by cryptix** [comment]
Yes, very much so and very glad you like the package! :)

From an API perspective, I really like [functional options](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

The gist of the link applied to this client API could be something like this:

```go
type Option func(*MyClient) error

func SetPort(i int) Option {
  // maybe do something with i first
  return func(c *MyClient) error {
    if i > 65536 { // validate i
       // .. oh well..
    }
    // can set private fields!
    c.port = i
  }
}

func NewClient(opts ...Option) (*MyClient,error) {
 var c MyClient
 // set defaults
 c.port = 1234
 // use passed options to override
  for _,o := range opts {
    if err:= o(&c); err!=nil {
      return nil, err
    }
  }
  return &c, nil
}
```

What options do you have in mind besides tunnel length?

**Comment by eyedeekay** [comment]
Well besides tunnel length, the common options are inbound/outbound length variance, tunnel quantity, and backup tunnel quantity. I'd like to start with those. I've been going over the i2cp options and most of the others don't seem to apply, but inbound.nickname seems like it could possibly be useful.

**Comment by cryptix** [comment]
I always wanted to add support for [encrypted leaseSets](https://geti2p.net/spec/proposals/121-encrypted-leaseset) (check out the superceded rfc also if you find it interesting).

You obviously don't have to add support for it all the way but it would be very nice if you keep the requirements for setting the required options in mind. IIRC those are normal i2cp options as well.

**Comment by eyedeekay** [comment]
From what I'm seeing I don't think there's any reason not to, so I'll try.

Unless you have any objections, instead of changing the behavior of NewClient(), I'm going to make a new function NewClientFromOptions() for passing the options to. That way existing applications don't need to change unless they make use of the options.

**Comment by cryptix** [comment]
I think this is mostly closed by #7.
